includes/RemovePIIJob.php

```diff
@@ -0,0 +1,482 @@
+				foreach ( $value as $name => $fields ) {
+					try {
+						$dbw->update(
+							$key,
+							$fields['fields'],
+							$fields['where'],
+							__METHOD__
+						);
+					} catch( Exception $e ) {
+						$this->setLastError( get_class( $e ) . ': ' . $e->getMessage() );
+
+						continue;
+					}
+				}
+			}
+		}
+
+		$logTitle = Title::newFromText( 'CentralAuth', NS_SPECIAL )->getSubpage( $userNewName );
+		$dbw->delete(
+			'logging', [
+				'log_action' => 'rename',
+				'log_title' => $logTitle->getDBkey(),
+				'log_type' => 'gblrename'
+			]
+		);
+
+		$dbw->delete(
+			'logging', [
+				'log_action' => 'renameuser',
+				'log_title' => $oldName->getTitleKey(),
+				'log_type' => 'renameuser'
+			]
+		);
+
+		$dbw->delete(
+			'recentchanges', [
+				'rc_log_action' => 'rename',
+				'rc_title' => $logTitle->getDBkey(),
+				'rc_log_type' => 'gblrename'
+			]
+		);
+
+		$dbw->delete(
+			'recentchanges', [
+				'rc_log_action' => 'renameuser',
+				'rc_title' => $oldName->getTitleKey(),
+				'rc_log_type' => 'renameuser'
+			]
+		);
+
+		$user = User::newSystemUser( 'MediaWiki default', [ 'steal' => true ] );
+
+		if ( !$user ) {
+			$this->setLastError( 'Invalid username' );
+
+			return false;
+		}
+
+		$userGroupManager = MediaWikiServices::getInstance()->getUserGroupManager();
+
+		// Hide deletions from RecentChanges
+		$userGroupManager->addUserToGroup( $user, 'bot', null, true );
+
+		$dbr = wfGetDB( DB_REPLICA, [], $this->database );
+		$userPageTitle = $oldName->getUserPage();
+
+		$namespaces = [
+			NS_USER,
+			NS_USER_TALK
+		];
+		
+		if ( class_exists( 'UserProfilePage' ) ) {
+			array_push( $namespaces,
+				NS_USER_WIKI,
+				NS_USER_WIKI_TALK,
+				NS_USER_PROFILE,
+				NS_USER_PROFILE_TALK
+			);
+		}
+		
+		if ( ExtensionRegistry::getInstance()->isLoaded( 'BlogPage' ) ) {
+			array_push( $namespaces,
+				NS_BLOG,
+				NS_BLOG_TALK
+			);
+		}
+
+		if ( ExtensionRegistry::getInstance()->isLoaded( 'SimpleBlogPage' ) ) {
+			array_push( $namespaces,
+				NS_USER_BLOG,
+				NS_USER_BLOG_TALK
+			);
+		}
+
+		$rows = $dbr->select(
+			'page', [
+				'page_namespace',
+				'page_title'
+			], [
+				'page_namespace IN (' . implode( ',', $namespaces ) . ')',
+				'(page_title ' . $dbr->buildLike( $userPageTitle->getDBkey() . '/', $dbr->anyString() ) .
+				' OR page_title = ' . $dbr->addQuotes( $userPageTitle->getDBkey() ) . ')'
+			],
+			__METHOD__
+		);
+
+		$dbw->query( 'DELETE FROM revision WHERE rev_id IN (SELECT rev_id FROM `revision` LEFT JOIN `page` ON rev_page = page_id WHERE' . '(page_title ' . $dbw->buildLike( $userPageTitle->getDBkey() . '/', $dbw->anyString() ) .
+				' OR page_title = ' . $dbw->addQuotes( $userPageTitle->getDBkey() ) . '))' );
+
+		$error = '';
+		foreach ( $rows as $row ) {
+			$title = Title::newFromRow( $row );
+
+			$userPage = WikiPage::factory( $title );
+			$status = @$userPage->doDeleteArticleReal( '', $user, true, null, $error, null, [], 'delete', true );
```

Comment:
Why are we suppressing warnings? That is an indication of a fault in the deletion logic...

If you need to suppress warnings, please use the best practices from https://github.com/wikimedia/at-ease.

Reply:
Because when the revision is deleted before the page it gives undefined index 'content' warning. And I can't do revision deletion after because then the page no longer exists in page table and the join doesn't work. But I can switch to atEase.

Reply:
Changed to atEase